### PR TITLE
Replace Scrivener with Flop

### PR DESCRIPTION
**Pull request by lawik**
This is pretty much a bug-by-bug compatible implementation as far as I can tell.

### CODE REVIEW

**Comment by lawik** [comment]
Takes care of issue #1656 

Flop wants the struct so it is a bit messy. There is a lot of potential tidying up that one could do. But this just rips out the old and provides identical functionality.

File: lib/nerves_hub/devices.ex [hunk 1]
```diff
@@ -70,10 +70,17 @@ defmodule NervesHub.Devices do
...
+    {entries, _pager} = get_devices_by_org_id_and_product_id_with_pager(org_id, product_id, opts)
+    entries
+  end
+
+  def get_devices_by_org_id_and_product_id_with_pager(org_id, product_id, opts) do
     pagination = Map.get(opts, :pagination, %{})
     sorting = Map.get(opts, :sort, {:asc, :identifier})
     filters = Map.get(opts, :filters, %{})
 
+    flop = %Flop{page: pagination[:page] || 1, page_size: pagination[:page_size] || 10}
```
Comment (nshoes) [inline lib/nerves_hub/devices.ex:82]:
```suggestion
    pagination = Map.get(opts, :pagination, %{page: 1, page_size: 10})
    sorting = Map.get(opts, :sort, {:asc, :identifier})
    filters = Map.get(opts, :filters, %{})

    flop = %Flop{page: pagination[:page], page_size: pagination[:page_size]}
```

File: lib/nerves_hub/devices.ex [hunk 2]
```diff
@@ -86,21 +93,24 @@ defmodule NervesHub.Devices do
...
+    |> Flop.run(flop)
   end
 
+  @spec filter(integer(), map()) :: {list(Device.t()), Flop.Meta.t()}
   def filter(product_id, opts) do
     pagination = Map.get(opts, :pagination, %{})
     sorting = Map.get(opts, :sort, {:asc, :identifier})
     filters = Map.get(opts, :filters, %{})
 
+    flop = %Flop{page: pagination.page, page_size: pagination.page_size}
```
Comment (nshoes) [inline lib/nerves_hub/devices.ex:105]: Do we care that these could be `nil` here? We could adopt my suggestion from above if so.

File: lib/nerves_hub/devices.ex [hunk 3]
```diff
@@ -70,10 +70,17 @@ defmodule NervesHub.Devices do
   end
 
   def get_devices_by_org_id_and_product_id(org_id, product_id, opts) do
+    {entries, _pager} = get_devices_by_org_id_and_product_id_with_pager(org_id, product_id, opts)
```
Comment (nshoes) [inline lib/nerves_hub/devices.ex:73]: I was originally off-put by `pager`, but its definition is "one who pages"... I love it 😆

File: lib/nerves_hub_web/views/layout_view.ex [hunk 4]
```diff
@@ -83,40 +83,76 @@ defmodule NervesHubWeb.LayoutView do
...
+        end_range: end_range,
+        distance: distance,
+        anchor: anchor
+      })
+
+    ~H"""
+    <div class="btn-group btn-group-toggle">
+      <div :if={@page_number > 1}>
+        <%= link("&lt;&lt;",
+          to: "?page=#{@page_number - 1}#{@anchor}",
```
Comment (nshoes) [inline lib/nerves_hub_web/views/layout_view.ex:102]: Are we resolving an index starting at 1 vs 0 here? Is Flop zero-based?

**Review by nshoes — CHANGES_REQUESTED** [review]
Thank you so much for doing this! Couple issues I found:

#### Cannot get back to page 1 after navigating to other pages
The double carets or clicking `1` doesn't send you back to the first page.

CleanShot 2024-11-21 at 07 43 28

#### Single page shows two page `1` icons
They both link to page 1, but it's confusing and should only display a single tile.
CleanShot 2024-11-21 at 07 46 05@2x

**Comment by nshoes** [comment]
> But this just rips out the old and provides identical functionality.

I didn't see this before making my review, this makes me think that the bugs I listed already existed, but I'd love to take care of them here.

**Comment by lawik** [comment]
@nshoes no worries, it is indeed bug-for-bug-compatible. I hope to take some time next week and make it not terrible. But I think replacing the dep is a good starting point and then improve from there.

File: lib/nerves_hub_web/views/layout_view.ex (hunk 4 above)
Comment (lawik) [reply lib/nerves_hub_web/views/layout_view.ex:102]: Offset/Limit mode is zero-based, page-mode is not. Fairly certain :)

File: lib/nerves_hub/devices.ex (hunk 3 above)
Comment (lawik) [reply lib/nerves_hub/devices.ex:73]: I didn't even think very hard. But of course intended this very fitting name :P

File: lib/nerves_hub/devices.ex (hunk 2 above)
Comment (lawik) [reply lib/nerves_hub/devices.ex:105]: They are only used in one place and will never be nil so I don't think we care.

**Review by nshoes — APPROVED** [review]

**Pull request merged**